### PR TITLE
ARROW-15110:  [C++][Gandiva] Revert change on Gandiva's cache policy

### PR DESCRIPTION
**Pull request by augustoasilva**
Revert change on "ARROW-12986: [C++][Gandiva] Implement new cache eviction policy algorithm in Gandiva"

This reverts commit [a628ee0](https://github.com/s1mbi0se/arrow/commit/a628ee0bd3e5644cc9dc1fa33935c0abf8355c90)

The commit that is being revered is causing random memory corruptions, based whe compiling the Gandiva to jars file for JNI:

```
#A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fdb8d427af8, pid=20974, tid=0x00007fdb8b669700
#
# JRE version: OpenJDK Runtime Environment (8.0_141-b16) (build 1.8.0_141-b16)
# Java VM: OpenJDK 64-Bit Server VM (25.141-b16 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libgandiva_jni.soa26ad92b-cf80-4777-98fc-5ea749c149ac+0x4afaf8]  gandiva::Projector::Make(std::shared_ptr<arrow::Schema>, std::vector<std::shared_ptr<gandiva::Expression>, std::allocator<std::shared_ptr<gandiva::Expression> > > const&, gandiva::SelectionVector::Mode, std::shared_ptr<gandiva::Configuration>, std::shared_ptr<gandiva::Projector>*)+0x858
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
```

So this is a temporary revert 'till the actual memory corruption is found and fixed.

### CODE REVIEW

File: cpp/src/gandiva/projector.cc [hunk 1]
```diff
@@ -18,14 +18,15 @@
 #include "gandiva/projector.h"
 
 #include <memory>
+#include <thread>
```
Comment (projjal) [inline cpp/src/gandiva/projector.cc:21]: why are these being included?

Comment (augustoasilva) [reply cpp/src/gandiva/projector.cc]: I did not saw it, so removed.

File: cpp/src/gandiva/gandiva_object_cache.cc [hunk 2]
```diff
@@ -17,34 +17,25 @@
...
+#include <utility>
+
 namespace gandiva {
 
 GandivaObjectCache::GandivaObjectCache(
     std::shared_ptr<Cache<ExpressionCacheKey, std::shared_ptr<llvm::MemoryBuffer>>>&
         cache,
     ExpressionCacheKey key)
-    : cache_key_(key) {
+    : cache_key_(std::move(key)) {
```
Comment (projjal) [inline cpp/src/gandiva/gandiva_object_cache.cc:28]: why this change?

File: cpp/src/gandiva/filter.cc [hunk 3]
```diff
@@ -18,13 +18,16 @@
 #include "gandiva/filter.h"
 
 #include <memory>
+#include <thread>
```
Comment (projjal) [inline cpp/src/gandiva/filter.cc:21]: These includes are also not necessary. Can you review all the includes in the PR.

Comment (augustoasilva) [reply cpp/src/gandiva/filter.cc]: Done!

File: cpp/src/gandiva/gandiva_object_cache.cc (hunk 2 above)
Comment (projjal) [reply cpp/src/gandiva/gandiva_object_cache.cc:28]: its ok here.

**Review by projjal — APPROVED** [review]
